google-cloud-storage/src/main/java/com/google/cloud/storage/UnifiedOpts.java

```diff
@@ -1920,6 +1935,184 @@ public Mapper<MoveObjectRequest.Builder> moveObject() {
+                  "Content-Transfer-Encoding",
+                  "Content-Type",
+                  "Date",
+                  "ETag",
+                  "If-Match",
+                  "If-None-Match",
+                  "Keep-Alive",
+                  "Range",
+                  "TE",
+                  "Trailer",
+                  "Transfer-Encoding",
+                  "User-Agent",
+                  "X-Goog-Api-Client",
+                  "X-Goog-Content-Length-Range",
+                  "X-Goog-Copy-Source-Encryption-Algorithm",
+                  "X-Goog-Copy-Source-Encryption-Key",
+                  "X-Goog-Copy-Source-Encryption-Key-Sha256",
+                  "X-Goog-Encryption-Algorithm",
+                  "X-Goog-Encryption-Key",
+                  "X-Goog-Encryption-Key-Sha256",
+                  "X-Goog-Gcs-Idempotency-Token",
+                  "X-Goog-Request-Params",
+                  "X-Goog-User-Project",
+                  "X-HTTP-Method-Override",
+                  "X-Upload-Content-Length",
+                  "X-Upload-Content-Type")
+              .map(Utils::headerNameToLowerCase)
+              .collect(ImmutableSet.toImmutableSet());
+
+      ImmutableSet<String> prefixes =
+          Stream.of("X-Goog-Meta-")
+              .map(Utils::headerNameToLowerCase)
+              .collect(ImmutableSet.toImmutableSet());
+
+      BLOCKLIST =
+          name -> {
+            if (fullHeaderNames.contains(name)) {
+              return true;
+            }
+
+            for (String prefix : prefixes) {
+              if (name.startsWith(prefix)) {
+                return true;
+              }
+            }
+            return false;
+          };
+    }
+
+    private Headers(ImmutableMap<String, String> val) {
+      super(StorageRpc.Option.EXTRA_HEADERS, val);
+    }
+
+    @Override
+    public Mapper<GrpcCallContext> getGrpcMetadataMapper() {
+      return ctx -> {
+        if (val.isEmpty()) {
+          return ctx;
+        }
+        Set<String> existingHeaderNames =
+            ctx.getExtraHeaders().keySet().stream()
+                .map(Utils::headerNameToLowerCase)
+                .collect(Collectors.toSet());
+        Map<String, List<String>> wrapped = new HashMap<>();
+        for (Entry<String, String> e : val.entrySet()) {
+          String key = Utils.headerNameToLowerCase(e.getKey());
+          if (existingHeaderNames.contains(key)) {
+            continue;
+          }
+          wrapped.put(key, ImmutableList.of(e.getValue()));
+        }
+        return ctx.withExtraHeaders(wrapped);
+      };
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Mapper<ImmutableMap.Builder<StorageRpc.Option, Object>> mapper() {
+      return optionBuilder -> {
+        if (val.isEmpty()) {
+          return optionBuilder;
+        }
+        // not ideal, but ImmutableMap.Builder doesn't have any read methods so we can detect
+        // collision
+        // before build time.
```

Comment:
nit: can probably condense this into 2 lines.

Reply:
Ah, the formatter wrapped the first line. I'll clean this up.